### PR TITLE
Bitcoin 0.15.1

### PR DESCRIPTION
**Pull request by luke-jr**
This is the first revamped ebuild form that doesn't use the eclass, so extra care should be taken in reviewing.

It's been live on the bitcoin overlay for ~10 days now, and there are no known outstanding issues.

For now, `linguas_sn` is defined in bitcoin-qt's metadata.xml. Should I move that to the global linguas?

### CODE REVIEW

**Comment by floppym** [comment]
> For now, linguas_sn is defined in bitcoin-qt's metadata.xml. Should I move that to the global linguas?

I would suggest removing the "linguas" USE flags altogether.

https://www.gentoo.org/support/news-items/2016-06-23-l10n-use_expand.html

If the build system uses LINGUAS as an environment variable, you can remove the USE flags and let it work automagically.

If  the ebuild needs USE flags, replace "linguas" with "l10n" and normalize the language codes to BCP-47. If there are BCP-47 values missing, I would be happy to merge patches for profiles/desc/l10n.desc.

**Comment by luke-jr** [comment]
Is there a sensible way to do that normalisation? Or just do something ugly like http://codepad.org/1lPHA9lX ?

**Comment by floppym** [comment]
> Is there a sensible way to do that normalisation?

I'm not aware of any existing helper code for that, sorry.

File: dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+SRC_URI="
+	https://github.com/${MyPN}/${MyPN}/archive/${BITCOINCORE_COMMITHASH}.tar.gz -> ${MyPN}-v${PV}.tar.gz
+	http://bitcoinknots.org/files/0.15.x/${KNOTS_PV}/${KNOTS_P}.patches.txz -> ${KNOTS_P}.patches.tar.xz
+"
+CORE_DESC="https://bitcoincore.org/en/2017/11/11/release-${PV}/"
+KNOTS_DESC="http://bitcoinknots.org/files/0.15.x/${KNOTS_PV}/${KNOTS_P}.desc.html"
+
+RDEPEND="
+	!libressl? ( dev-libs/openssl:0[-bindist] ) libressl? ( dev-libs/libressl )
```
Comment (SoapGentoo) [inline dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild]: on separate lines pls, and both `openssl` and `libressl` require `:0=`

File: dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	>=dev-libs/libsecp256k1-0.0.0_pre20151118[recovery]
+	dev-libs/univalue
+	>=dev-libs/boost-1.52.0[threads(+)]
```
Comment (SoapGentoo) [inline dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild]: `:=`

File: dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+"
+DEPEND="${RDEPEND}
+	>=app-shells/bash-4.1
```
Comment (SoapGentoo) [inline dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild]: this is already part of EAPI 6, not required

File: dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+"
+
+DOCS="doc/bips.md doc/release-notes.md"
```
Comment (SoapGentoo) [inline dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild]: array please

File: dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+	if use knots; then
+		eapply "$(KNOTS_PATCH f)"
+		eapply "$(KNOTS_PATCH branding)"
+		eapply "$(KNOTS_PATCH ts)"
+	fi
+
+	eapply_user
+
+	echo '#!/bin/true' >share/genbuild.sh
```
Comment (SoapGentoo) [inline dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild]: missing `|| die`

File: dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	mkdir -p src/obj
```
Comment (SoapGentoo) [inline dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild]: missing `|| die`

File: dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	echo "#define BUILD_SUFFIX gentoo${PVR#${PV}}" >src/obj/build.h
```
Comment (SoapGentoo) [inline dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild]: missing `|| die`

File: dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+		--disable-experimental-asm
+		--without-qtdbus
+		--without-libevent
+		--without-qrencode
+		--without-miniupnpc
+		--disable-tests
+		--disable-wallet
+		--disable-zmq
+		--enable-util-tx
+		--disable-util-cli --disable-bench --without-libs --without-daemon --without-gui
```
Comment (SoapGentoo) [inline dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild]: one option per line please

File: net-libs/libbitcoinconsensus/libbitcoinconsensus-0.15.1.ebuild [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+SRC_URI="
+	https://github.com/${MyPN}/${MyPN}/archive/${BITCOINCORE_COMMITHASH}.tar.gz -> ${MyPN}-v${PV}.tar.gz
+	http://bitcoinknots.org/files/0.15.x/${KNOTS_PV}/${KNOTS_P}.patches.txz -> ${KNOTS_P}.patches.tar.xz
+"
+CORE_DESC="https://bitcoincore.org/en/2017/11/11/release-${PV}/"
+KNOTS_DESC="http://bitcoinknots.org/files/0.15.x/${KNOTS_PV}/${KNOTS_P}.desc.html"
+
+RDEPEND="
+	!libressl? ( dev-libs/openssl:0[-bindist] ) libressl? ( dev-libs/libressl )
```
Comment (SoapGentoo) [inline net-libs/libbitcoinconsensus/libbitcoinconsensus-0.15.1.ebuild]: same here

File: net-libs/libbitcoinconsensus/libbitcoinconsensus-0.15.1.ebuild [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+		eapply "$(KNOTS_PATCH f)"
+		eapply "$(KNOTS_PATCH branding)"
+		eapply "$(KNOTS_PATCH ts)"
+	fi
+
+	eapply_user
+
+	echo '#!/bin/true' >share/genbuild.sh
+	mkdir -p src/obj
+	echo "#define BUILD_SUFFIX gentoo${PVR#${PV}}" >src/obj/build.h
```
Comment (SoapGentoo) [inline net-libs/libbitcoinconsensus/libbitcoinconsensus-0.15.1.ebuild]: ditto

File: net-libs/libbitcoinconsensus/libbitcoinconsensus-0.15.1.ebuild [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+		--disable-ccache --disable-static
+		--with-system-libsecp256k1
+	)
+	econf "${my_econf[@]}"
+}
+
+src_install() {
+	default
+
+	rm "${D}/usr/bin/test_bitcoin"
```
Comment (SoapGentoo) [inline net-libs/libbitcoinconsensus/libbitcoinconsensus-0.15.1.ebuild]: missing `|| die` and use `${ED%/}` over `${D}`

File: net-p2p/bitcoin-qt/bitcoin-qt-0.15.1.ebuild [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+	>=dev-libs/boost-1.52.0[threads(+)]
+	upnp? ( >=net-libs/miniupnpc-1.9.20150916 )
+	wallet? ( sys-libs/db:$(db_ver_to_slot "${DB_VER}")[cxx] )
+	zeromq? ( net-libs/zeromq )
+	virtual/bitcoin-leveldb
+	bip70? ( dev-libs/protobuf )
+	qrcode? (
+		media-gfx/qrencode
+	)
+	dev-qt/qtgui:5 dev-qt/qtnetwork:5 dev-qt/qtwidgets:5
```
Comment (SoapGentoo) [inline net-p2p/bitcoin-qt/bitcoin-qt-0.15.1.ebuild]: again, one dep per line

File: dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild (hunk 1 above)
Comment (jonasstein) [reply dev-util/bitcoin-tx/bitcoin-tx-0.15.1.ebuild]: Here is the link to the on-dep-per line 
https://devmanual.gentoo.org/general-concepts/dependencies/

File: net-p2p/bitcoin-qt/bitcoin-qt-0.15.1.ebuild (hunk 12 above)
Comment (a17r) [reply net-p2p/bitcoin-qt/bitcoin-qt-0.15.1.ebuild]: also, probably needs `qtcore:5` as well.

File: net-p2p/bitcoin-qt/bitcoin-qt-0.15.1.ebuild [hunk 13]
```diff
@@ -0,0 +1,246 @@
+# Copyright 2010-2017 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
+
+DB_VER="4.8"
+inherit autotools bash-completion-r1 db-use gnome2-utils kde4-functions xdg-utils
```
Comment (a17r) [inline net-p2p/bitcoin-qt/bitcoin-qt-0.15.1.ebuild]: what's `kde4-functions` still doing there?

Comment (luke-jr) [reply net-p2p/bitcoin-qt/bitcoin-qt-0.15.1.ebuild]: For `buildsycoca` - do we not need that anymore? Not entirely sure what it does...

File: net-libs/libbitcoinconsensus/libbitcoinconsensus-0.15.1.ebuild (hunk 11 above)
Comment (luke-jr) [reply net-libs/libbitcoinconsensus/libbitcoinconsensus-0.15.1.ebuild]: Can't die, since it won't always exist.

**Comment by luke-jr** [comment]
Completed the other changes.

File: net-p2p/bitcoin-qt/bitcoin-qt-0.15.1.ebuild (hunk 13 above)
Comment (a17r) [reply net-p2p/bitcoin-qt/bitcoin-qt-0.15.1.ebuild]: not in a long time, kded is doing that as-needed. the only thing left in `buildsycoca` is a permissions check.

Comment (luke-jr) [reply net-p2p/bitcoin-qt/bitcoin-qt-0.15.1.ebuild]: Okay, dropped it.

File: net-libs/libbitcoinconsensus/libbitcoinconsensus-0.15.1.ebuild (hunk 11 above)
Comment (SoapGentoo) [reply net-libs/libbitcoinconsensus/libbitcoinconsensus-0.15.1.ebuild]: not a reason, use `-f` then

**Comment by luke-jr** [comment]
Also rebased to resolve merge conflict (leaving it pushing overnight as my ISP sucks)

File: net-p2p/bitcoind/bitcoind-0.15.1.ebuild [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+	>=dev-libs/libsecp256k1-0.0.0_pre20151118[recovery]
+	dev-libs/univalue
+	>=dev-libs/boost-1.52.0:=[threads(+)]
+	upnp? ( >=net-libs/miniupnpc-1.9.20150916 )
+	wallet? ( sys-libs/db:$(db_ver_to_slot "${DB_VER}")[cxx] )
+	zeromq? ( net-libs/zeromq )
+	virtual/bitcoin-leveldb
+"
+DEPEND="${RDEPEND}
+	sys-apps/sed
```
Comment (SoapGentoo) [inline net-p2p/bitcoind/bitcoind-0.15.1.ebuild]: remove part of @base

File: net-p2p/bitcoind/bitcoind-0.15.1.ebuild [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+	rm -r src/leveldb src/secp256k1 || die
+}
+
+src_configure() {
+	local my_econf=(
+		$(use_enable asm experimental-asm)
+		--without-qtdbus
+		--with-libevent
+		--without-qrencode
+		$(use_with upnp miniupnpc) $(use_enable upnp upnp-default)
```
Comment (SoapGentoo) [inline net-p2p/bitcoind/bitcoind-0.15.1.ebuild]: separate per line

**Comment by SoapGentoo** [comment]
@luke-jr you obviously need to split the commit up per package and use correct COMMITMSGs

**Comment by luke-jr** [comment]
Done

**Comment by jonasstein** [comment]
We can add now 
Closes: https://bugs.gentoo.org/640674
to the commit message on merge.

**Comment by luke-jr** [comment]
Populated the `Closes` headers in the commit messages.

**Comment by puchu** [comment]
any news on this? when will this get pushed to the main gentoo tree? the latest version in the portage tree is bitcoin-qt-0.13.2 which was released on the 3rd of Jan 2017

**Comment by puchu** [comment]
please also use the tarballs from https://bitcoin.org/bin/ see https://security.stackexchange.com/questions/124320/how-can-i-check-the-integrity-of-a-bitcoin-tarball-with-gpg as tarballs generated by github don't provide signed checksum files (and the checksum can change over time, see for example https://github.com/easybuilders/easybuild-easyconfigs/issues/5151). it think using gpg to verifiy that signatures is a "must have" not a "nice to have" feature

**Comment by puchu** [comment]
see https://github.com/puchu/gentoo/commit/a45bdce3f313d6274045f2eda8df4ce90b2a0923 . as the official tarball doesnt include the contrib/debian folder i added the files to the files dir

**Comment by luke-jr** [comment]
The tarballs on bitcoin.org are not maintained, and are (as you noticed) missing files. Your modified ebuild doesn't really work either.

Portage has its own Manifest files for verifying the integrity of the downloaded source code. That's more than enough - the weak link is already on syncing the portage tree anyway, since an attacker can modify the ebuild in that scenario.

**Pull request closed without merging**